### PR TITLE
Add restful api support.

### PR DESCRIPTION
**Pull request by qiuyesuifeng**

### CODE REVIEW

**Comment by qiuyesuifeng** [comment]
@siddontang @shenli @disksing PTAL

File: server/cluster.go [hunk 1]
```diff
@@ -410,3 +424,29 @@ func (c *raftCluster) putConfig(meta *metapb.Cluster) error {
...
+
+// GetScore gets store score from balancer.
+func (c *RaftCluster) GetScore(store *metapb.Store, status *StoreStatus) int {
+	storeInfo := &storeInfo{
+		store: store,
+		stats: status,
+	}
+
+	regionCount := c.cachedCluster.regions.regionCount()
+	return c.balancerWorker.balancer.(*resourceBalancer).score(storeInfo, regionCount)
```
Comment (siddontang) [inline server/cluster.go]: can we support a store score function in balancer worker. 

`c.balancerWorker.balancer.(*resourceBalancer).score` seems too tricky.

File: server/cache.go [hunk 2]
```diff
@@ -302,21 +329,53 @@ func (r *regionsInfo) heartbeatVersion(region *metapb.Region) (bool, *metapb.Reg
...
+			return &pdpb.ChangePeer{
+				ChangeType: raftpb.ConfChangeType_RemoveNode.Enum(),
+				Peer:       peer,
+			}
+		}
+	}
+
+	// Get add peer.
+	for _, peer := range curRegion.GetPeers() {
+		// Current region have region peer, but region does not have,
```
Comment (siddontang) [inline server/cache.go]: has

**Comment by disksing** [comment]
LGTM.

File: server/operator.go [hunk 3]
```diff
@@ -241,45 +273,68 @@ func (tlo *transferLeaderOperator) check(region *metapb.Region, leader *metapb.P
...
 	}
 
-	log.Infof("balance [%s], try to transfer leader from %s to %s", region, tlo.oldLeader, tlo.newLeader)
+	log.Infof("balance [%d][%s], try to transfer leader from %s to %s", tlo.Count, region, tlo.OldLeader, tlo.NewLeader)
 	return false, nil
 }
 
 // Do implements Operator.Do interface.
-func (tlo *transferLeaderOperator) Do(region *metapb.Region, leader *metapb.Peer) (bool, *pdpb.RegionHeartbeatResponse, error) {
+func (tlo *transferLeaderOperator) Do(region *metapb.Region, leader *metapb.Peer, scb callback, ecb callback) (bool, *pdpb.RegionHeartbeatResponse, error) {
```
Comment (siddontang) [inline server/operator.go]: what scb and ecb mean?

File: server/operator.go [hunk 4]
```diff
@@ -25,69 +25,77 @@ import (
 	"github.com/pingcap/kvproto/pkg/raftpb"
 )
 
-var balancerIDs uint64
+var ids uint64
```
Comment (siddontang) [inline server/operator.go]: ids is for multi ids, not proper here.

File: server/operator.go (hunk 3 above)
Comment (qiuyesuifeng) [inline server/operator.go]: `scb` means `start event callback`
`ecb` means `end event callback`

Comment (siddontang) [inline server/operator.go]: I prefer onStart and onEnd

File: server/region_cache.go [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+		}
+	}
+}
+
+type lruCache struct {
+	sync.RWMutex
+
+	// maxCount is the maximum number of items.
+	// 0 means no limit.
+	maxCount int
```
Comment (siddontang) [inline server/region_cache.go]: do we really need no limit cache?

Comment (qiuyesuifeng) [inline server/region_cache.go]: Not needed, actually.

File: server/cluster.go [hunk 6]
```diff
@@ -410,3 +423,29 @@ func (c *raftCluster) putConfig(meta *metapb.Cluster) error {
...
+
+// GetScore gets store score from balancer.
+func (c *RaftCluster) GetScore(store *metapb.Store, status *StoreStatus) int {
+	storeInfo := &storeInfo{
+		store: store,
+		stats: status,
+	}
+
+	regionCount := c.cachedCluster.regions.regionCount()
+	return c.balancerWorker.balancer.score(storeInfo, regionCount)
```
Comment (siddontang) [inline server/cluster.go]: can we supply a StoreScore function in balancerWorker?

**Comment by siddontang** [comment]
LGTM

**Pull request merged**